File: src/dotnet/FilterIcs/Feeds.cs
namespace FilterIcs;

public sealed record FeedDefinition(string Basename, string OutputName, string OutputFile, Allowlist Allowlist);

public sealed class RunResult
{
    public int Total { get; init; }
    public List<string> Written { get; } = new();
    public List<(string basename, string reason)> Skipped { get; } = new();
    public List<string> Removed { get; } = new();
    public long SourceBytes { get; init; }
    public Dictionary<string, IReadOnlyList<string>> Unmatched { get; } = new();
}

/// <summary>Multi-feed orchestration: fetch once, filter per file, reconcile outputs.</summary>
public static class Feeds
{
    public static (List<FeedDefinition> feeds, List<(string, string)> errors, List<string> discovered)
        LoadFeeds(string allowlistsDir)
    {
        var feeds = new List<FeedDefinition>();
        var errors = new List<(string, string)>();
        var discovered = new List<string>();

        foreach (string path in Directory.GetFiles(allowlistsDir, "*.json").OrderBy(p => p, StringComparer.Ordinal))
        {
            string basename = Path.GetFileName(path);
            discovered.Add(basename);
            try
            {
                var (fileName, names) = FeedFile.Load(path);
                string outputName = Naming.SanitizeOutputName(fileName);
                Allowlist allowlist = Allowlist.FromNames(names);
                feeds.Add(new FeedDefinition(basename, outputName, Naming.OutputFileName(outputName), allowlist));
            }
            catch (FeedException ex)
            {
                errors.Add((basename, ex.Message));
            }
        }
        return (feeds, errors, discovered);
    }

    public static RunResult Run(string sourceUrl, string allowlistsDir, string outputDir, Func<string, string>? fetcher = null)
    {
        fetcher ??= url => Fetch.FetchCalendarAsync(url).GetAwaiter().GetResult();

        var (feeds, errors, discovered) = LoadFeeds(allowlistsDir);

        // Duplicate output names → every colliding file is an error (FR-007).
        var byOutput = feeds.GroupBy(f => f.OutputFile);
        var valid = new List<FeedDefinition>();
        foreach (var group in byOutput)
        {
            var members = group.ToList();
            if (members.Count > 1)
                foreach (var f in members)
                    errors.Add((f.Basename, $"duplicate output name '{group.Key}'"));
            else
                valid.Add(members[0]);
        }

        // Fetch ONCE; parse-check the source once (throws ParseException on invalid source).
        string raw = fetcher(sourceUrl);
        Filter.FilterCalendar(raw, Allowlist.FromNames(Array.Empty<string>()));

        Directory.CreateDirectory(outputDir);
        string manifestPath = Path.Combine(outputDir, ".feeds.json");
        var old = Manifest.Read(manifestPath);
        var updated = new Dictionary<string, string>();
        var result = new RunResult
        {
            Total = discovered.Count,
            SourceBytes = System.Text.Encoding.UTF8.GetByteCount(raw),
        };

        foreach (var feed in valid)
        {
            FilterResult filtered = Filter.FilterCalendar(raw, feed.Allowlist);
            Render.AtomicWrite(Path.Combine(outputDir, feed.OutputFile), Render.Serialize(filtered.Calendar));
            result.Written.Add(feed.OutputName);
            updated[feed.Basename] = feed.OutputName;
            if (filtered.UnmatchedNames.Count > 0)
                result.Unmatched[feed.OutputName] = filtered.UnmatchedNames;
        }

        // Errored files: record; carry over last-good (keep manifest entry, leave file untouched).
        foreach (var (basename, reason) in errors)
        {
            result.Skipped.Add((basename, reason));
            if (old.ContainsKey(basename) && !updated.ContainsKey(basename))
                updated[basename] = old[basename];
        }

        // Reconcile removals/renames: any previously published output not produced this run is deleted.
        var produced = new HashSet<string>(updated.Values);
        foreach (string oldName in old.Values.Distinct().OrderBy(n => n, StringComparer.Ordinal))
        {
            if (!produced.Contains(oldName))
            {
                string target = Path.Combine(outputDir, Naming.OutputFileName(oldName));
                if (File.Exists(target))
                    File.Delete(target);
                result.Removed.Add(oldName);
            }
        }

        Manifest.Write(manifestPath, updated);
        return result;
    }
}

File: src/python/filter_ics/feeds.py
"""Multi-feed orchestration: fetch once, filter per file, reconcile published outputs.

See contracts/cli.md and contracts/publishing.md.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from pathlib import Path

from .allowlist import Allowlist, FeedError, build_allowlist, load_feed_file
from .fetch import fetch_calendar
from .filter import filter_calendar
from .manifest import read_manifest, write_manifest
from .naming import output_filename, sanitize_output_name
from .render import atomic_write, render


@dataclass(frozen=True)
class FeedDefinition:
    basename: str  # e.g. "engineering.json" — manifest key
    output_name: str  # sanitized base name, no extension
    output_file: str  # "<output_name>.ics"
    allowlist: Allowlist


@dataclass
class RunResult:
    total: int = 0  # allowlist files discovered
    written: list[str] = field(default_factory=list)  # output names written/updated
    skipped: list[tuple[str, str]] = field(default_factory=list)  # (basename, reason)
    removed: list[str] = field(default_factory=list)  # output names deleted
    source_bytes: int = 0
    unmatched: dict[str, list[str]] = field(default_factory=dict)  # output_name -> names


def load_feeds(allowlists_dir: str | Path) -> tuple[list[FeedDefinition], list[tuple[str, str]], list[str]]:
    """Parse every ``*.json`` in the folder. Returns (valid feeds, errors, discovered basenames)."""
    directory = Path(allowlists_dir)
    feeds: list[FeedDefinition] = []
    errors: list[tuple[str, str]] = []
    discovered: list[str] = []
    for path in sorted(directory.glob("*.json")):
        discovered.append(path.name)
        try:
            file_name, names = load_feed_file(path)
            output_name = sanitize_output_name(file_name)
            allowlist = build_allowlist(names)
        except FeedError as exc:
            errors.append((path.name, str(exc)))
            continue
        feeds.append(FeedDefinition(path.name, output_name, output_filename(output_name), allowlist))
    return feeds, errors, discovered


def run(
    source_url: str,
    allowlists_dir: str | Path,
    output_dir: str | Path,
    *,
    fetcher=fetch_calendar,
) -> RunResult:
    """Produce/refresh all feeds from one source fetch and reconcile the output directory.

    Raises FetchError (source fetch) or ParseError (source not iCalendar) — both map to a
    whole-run failure that writes nothing. Per-feed problems are recorded in the result and
    never abort the others.
    """
    feeds, errors, discovered = load_feeds(allowlists_dir)

    # Duplicate output names → every colliding file is an error (FR-007).
    by_output: dict[str, list[FeedDefinition]] = {}
    for feed in feeds:
        by_output.setdefault(feed.output_file, []).append(feed)
    valid: list[FeedDefinition] = []
    for output_file, group in by_output.items():
        if len(group) > 1:
            for feed in group:
                errors.append((feed.basename, f"duplicate output name '{output_file}'"))
        else:
            valid.append(group[0])

    # Fetch ONCE for all feeds; parse-check once (raises ParseError on invalid source).
    raw = fetcher(source_url)
    filter_calendar(raw, build_allowlist([]))  # validates the source parses

    out_dir = Path(output_dir)
    out_dir.mkdir(parents=True, exist_ok=True)
    manifest_path = out_dir / ".feeds.json"
    old = read_manifest(manifest_path)
    new: dict[str, str] = {}
    result = RunResult(total=len(discovered), source_bytes=len(raw))

    for feed in valid:
        filtered = filter_calendar(raw, feed.allowlist)
        atomic_write(out_dir / feed.output_file, render(filtered.calendar))
        result.written.append(feed.output_name)
        new[feed.basename] = feed.output_name
        if filtered.unmatched_names:
            result.unmatched[feed.output_name] = filtered.unmatched_names

    # Errored files: record; carry over last-good (keep manifest entry, leave file untouched).
    for basename, reason in errors:
        result.skipped.append((basename, reason))
        if basename in old and basename not in new:
            new[basename] = old[basename]

    # Reconcile removals/renames: any previously published output not produced this run is deleted.
    produced = set(new.values())
    for old_name in sorted(set(old.values())):
        if old_name not in produced:
            target = out_dir / output_filename(old_name)
            if target.exists():
                target.unlink()
            result.removed.append(old_name)

    write_manifest(manifest_path, new)
    return result
